sympy/sets/sets.py

```diff
@@ -1525,6 +1504,21 @@ def as_relational(self, symbol):
 
         return And(A_rel, B_rel)
 
+    @property
+    def is_iterable(self):
+        if self.args[0].is_iterable:
+            return True
+
+    def __iter__(self):
+        if not self.is_iterable:
+            raise TypeError("{} is not iterable".format(self))
+        A, B = self.args
+        for a in A:
+            if a not in B:
+                    yield a
```

Comment:
I think this indicates a general problem here. The `not in` will raise if the element is possibly in the set:
```julia
In [2]: 1 in FiniteSet(x, y)                                                                                                      
---------------------------------------------------------------------------
TypeError
```
What should iteration return for something like:
```julia
In [6]: S({x, y}) - S({z})                                                                                                        
Out[6]: {x, y} \ {z}
```
we don't know if either x or y is in the set...

Reply:
I made `__iter__` to be a programmatic interface, along with `__contains__`.
And `__iter__` should also fail if `__contains__` fails.

Reply:
Okay so raising is the intention. Fair enough